src/coreclr/jit/fgbasic.cpp

```diff
@@ -516,6 +562,155 @@ void Compiler::fgReplaceSwitchJumpTarget(BasicBlock* blockSwitch, BasicBlock* ne
+//
+void Compiler::fgReplaceEhfSuccessor(BasicBlock* block, BasicBlock* oldSucc, BasicBlock* newSucc)
+{
+    assert(block != nullptr);
+    assert(oldSucc != nullptr);
+    assert(newSucc != nullptr);
+    assert(block->KindIs(BBJ_EHFINALLYRET));
+    assert(fgPredsComputed);
+
+    BBehfDesc* const ehfDesc   = block->GetJumpEhf();
+    unsigned         succCount = ehfDesc->bbeCount;
+    BasicBlock**     succTab   = ehfDesc->bbeSuccs;
+
+    unsigned i = 0;
+
+    // Walk the successor table looking for blocks to update the preds for
+    while (i < succCount)
+    {
+        if (succTab[i] == oldSucc) // We will update when succTab[i] matches
+        {
+            // Remove the old edge [block => oldSucc]
+            //
+            fgRemoveAllRefPreds(oldSucc, block);
+
+            // Change the succTab entry to branch to the new location
+            //
+            succTab[i] = newSucc;
+
+            // Create the new edge [block => newSucc]
+            //
+            FlowEdge* const newEdge = fgAddRefPred(newSucc, block);
+
+            // Now set the correct value of newEdge's DupCount
+            // and replace any other successors in succTab[] that go to oldSucc.
+            //
+            i++;
+            while (i < succCount)
+            {
+                if (succTab[i] == oldSucc)
+                {
+                    // We also must update this entry in the succTab
+                    //
+                    succTab[i] = newSucc;
+                    newSucc->bbRefs++;
+                    newEdge->incrementDupCount();
+                }
+                i++; // Check the next entry in succTab[]
+            }
+            return;
+        }
+        i++; // Check the next entry in succTab[] for a match
+    }
+    noway_assert(!"Did not find oldTarget in succTab[]");
+}
+
+//------------------------------------------------------------------------
+// fgRemoveEhfSuccessor: update BBJ_EHFINALLYRET block to remove `succ` as a successor.
+// Updates the predecessor list of `succ`.
+//
+// Arguments:
+//   block   - BBJ_EHFINALLYRET block
+//   succ    - successor
+//
+void Compiler::fgRemoveEhfSuccessor(BasicBlock* block, BasicBlock* succ)
+{
+    assert(block != nullptr);
+    assert(succ != nullptr);
+    assert(fgPredsComputed);
+    assert(block->KindIs(BBJ_EHFINALLYRET));
+
+    // Don't `assert(succ->isBBCallAlwaysPairTail())`; we've already unlinked the CALLFINALLY
+    assert(succ->KindIs(BBJ_ALWAYS));
+
+    fgRemoveRefPred(succ, block);
+
+    BBehfDesc* const ehfDesc   = block->GetJumpEhf();
+    unsigned         succCount = ehfDesc->bbeCount;
+    BasicBlock**     succTab   = ehfDesc->bbeSuccs;
+    bool             found     = false;
+
+    // Walk the successor table looking for the specified successor block.
+    for (unsigned i = 0; i < succCount; i++)
+    {
+        if (succTab[i] == succ)
+        {
+            // If it's not the last one, move everything after in the table down one slot.
```

Comment:
Maybe `assert(!found)` here instead of the loop below?

Reply:
What I meant to do was have a `break;` after the debug loop, since I don't want to keep iterating in the non-debug case.